### ISSUE TITLE
Feature request: Initial embeddings

### ISSUE THREAD

**Issue by mingruimingrui** [facebookresearch/PyTorch-BigGraph#17]
Having a feature to allow the user to define initial embeddings easily would be very helpful!
Currently the only easy work-about is through modifying checkpoint files (and using init_path) but less convoluted methods are not yet available.

For streaming behaviour data (think of reddit and twitter), it's meaningful to allow embeddings to change a little but also not too much from the previous training cycle. Also new data needs to be accommodated.

It would be cool if train.py could accept custom initial embeddings and create embeddings for new entities in edge_list.

Thanks FAIR for open sourcing this awesome tool!

**Comment by lerks** [comment]
I think PBG does more or less what you're asking for. Imagine you have a sequence of edgelists and want to train on one at a time, by bootstrapping the embeddings for an edgelist with the embeddings learned on the previous edgelist. Then, on the very first one, you leave `init_path` unset, which means embeddings are initialized randomly. When training on the second one, you set `init_path` to the `checkpoint_path` of the previous run (and you set `checkpoint_path` to a new empty directory, otherwise PBG would think that training is already complete and do nothing). Then you continue like this until you covered all edgelist files. This should work fine. Note however that, apart from the edgelist path, the configs of all these runs must be identical, including the entity counts.

The TSV importer does support this use-case quite well too, as you can pass to it an arbitrary number of TSV edgelist files and it will partition all their entities and consistently bucket all the edgelists.

Problems may arise if you don't yet have all edgelists available when you start training on the first one. (I'm guessing this is what you meant by "streaming behavior"). In that case you may not be aware at the beginning of entities that will only start appearing at a later time. Off the top of my head there are two adjustments that need to be made for this to work:
- The `import_from_tsv` script needs to be updated to support reading a previously-produced dictionary file and use it to initialize the assignments from entities to partitions + offsets. (Observe that while the assignment of an entity to a partition should be uniformly at random, the offsets within a partition don't need to be shuffled, so it's fine to just append new entities at the end).
- There needs to be a script that allows to "enlarge" a checkpoint, i.e., take a `.h5` file containing the embeddings of a partition of N entities and produce one for N+M entities, by copying over the old embeddings and initializing the new ones with random data.

**Comment by adamlerer** [comment]
Yes, this is a common use case. @lerks I think we also want to make it easy to initialize from an *externally produced* set of initial embeddings for some of the entities, not just from an existing PBG checkpoint. E.g. it might be useful to initialize with a (properly normalized) SVD of the adjacency matrix or initial embeddings from some smaller-scale embeddings package.

I think we need to add additional support in tsv_to_checkpoint to also take in a tsv file in the embeddings output format (i.e. each row contains entity name followed by tab-separated embeddings) and automatically create an initial checkpoint. This would technically support both use cases (initial embeddings from PBG / initial embeddings from elsewhere). I don't think it's as important to support adding entities to a checkpoint "in place" (although I could be wrong).

**Comment by mingruimingrui** [comment]
> * The `import_from_tsv` script needs to be updated to support reading a previously-produced dictionary file and use it to initialize the assignments from entities to partitions + offsets. (Observe that while the assignment of an entity to a partition should be uniformly at random, the offsets within a partition don't need to be shuffled, so it's fine to just append new entities at the end).
> * There needs to be a script that allows to "enlarge" a checkpoint, i.e., take a `.h5` file containing the embeddings of a partition of N entities and produce one for N+M entities, by copying over the old embeddings and initializing the new ones with random data.

Those are exactly what I'm looking for! Certainly makes the package more robust and simple to use.

I apologize for the loose use of vocabulary in the previous post. I meant to refer to datasets which regularly receives new entries (and therefore streaming). User behaviour data are often like that.

**Comment by mingruimingrui** [comment]
Hi @lerks let me check with you real quick. Help me catch the errors in the plan

`init_path` should be a folder containing the`embeddings_{entity}_{part}.v{version}.h5` files. `version` will use contents of `checkpoint_version.txt` if present in `init_path` else version will be assumed to be 0, meaning `embeddings_{entity}_{part}.v0.h5` is perfectly valid.

Also `embeddings.h5` can contain only embeddings and a zeroed optimiser state will be assumed.

**Comment by lerks** [comment]
All you said is correct.

If the `.h5` file contains no dataset named `optimizer/state_dict` then the optimizer will be initialized with its default values.
